src/scala/com/github/johnynek/bazel_deps/MakeDeps.scala

```diff
@@ -14,122 +16,130 @@ object MakeDeps {
+  private def fromE[A](ioa: IO[Either[Throwable, A]], failCode: ExitCode, msg: => String): EitherT[IO, ExitCode, A] =
+    fromEither(ioa) { err =>
+      IO.blocking {
+        logger.error(msg, err)
+        failCode
+      }
     }
 
-    val parser = if (g.depsFile.endsWith(".json")) new JawnParser else Yaml
-    val model = Decoders.decodeModel(parser, content) match {
-      case Right(m) => m
-      case Left(err) =>
-        logger.error(s"Failed to parse ${g.depsFile}.", err)
-        System.exit(1)
-        sys.error("unreachable")
-    }
-    val projectRoot = g.repoRoot
+  private def fromEither[E, A](ioa: IO[Either[E, A]])(fn: E => IO[ExitCode]): EitherT[IO, ExitCode, A] =
+    EitherT(ioa.flatMap {
+      case Right(a) => IO.pure(Right(a))
+      case Left(err) => fn(err).map(Left(_))
+    })
 
-    resolverCachePath(model, projectRoot).flatMap(runResolve(projectRoot, model, _)) match {
-      case Failure(err) =>
-        logger.error("resolution and sha collection failed", err)
-        System.exit(1)
-      case Success((normalized, shas, duplicates)) =>
-        // build the BUILDs in thirdParty
-        val targets = Writer.targets(normalized, model) match {
-          case Right(t) => t
-          case Left(errs) =>
-            errs.toList.foreach { e => logger.error(e.message) }
-            System.exit(-1)
-            sys.error("exited already")
-        }
-        val formatter: Writer.BuildFileFormatter = g.buildifier match {
-          // If buildifier is provided, run it with the unformatted contents on its stdin; it will print the formatted
-          // result on stdout.
-          case Some(buildifierPath) => (p, s) => {
-            val output = new java.lang.StringBuilder()
-            val error = new java.lang.StringBuilder()
-            val processIO = new ProcessIO(
-              os => {
-                os.write(s.getBytes(FS.charset))
-                os.close()
-              },
-              BasicIO.processFully(output),
-              BasicIO.processFully(error)
-            )
-            val exit = Process(List(buildifierPath, "-path", p.asString, "-"), projectRoot.toFile).run(processIO).exitValue
-            // Blocks until the process exits.
-            if (exit != 0) {
-              logger.error(s"buildifier $buildifierPath failed (code $exit) for ${p.asString}:\n$error")
-              System.exit(-1)
-              sys.error("unreachable")
-            }
-            output.toString
+  private def runFS[A](fsIO: FS.Result[A], artifactsSize: Int, projectRoot: Path, checkOnly: Boolean): EitherT[IO, ExitCode, ExitCode] =
+    EitherT(IO.blocking { if (checkOnly) {
+      val check = new FS.ReadCheckExec(projectRoot)
+      fsIO.foldMap(check) match {
+        case Failure(err) =>
+          logger.error("Failure during IO:", err)
+          val errs = check.logErrorCount { ce => logger.error(ce.message) }
+          logger.error(s"found $errs errors.")
+          Left(ExitCode(-1))
```

Comment:
It was this behavior prior to this PR too but exit codes are a number between 0 and 255, or modulo 255, so -1 might be misleading.